### ISSUE TITLE
DRBD 8.4 monitoring not working

### ISSUE THREAD

**Issue by vvidic** [ClusterLabs/ha_cluster_exporter#117]
Required `--json` option does not work with DRBD 8.4 kernel module loaded:

```
$ strace -e trace=execve /sbin/drbdsetup status --json
execve("/sbin/drbdsetup", ["/sbin/drbdsetup", "status", "--json"], 0x7fff0f266ca0 /* 40 vars */) = 0
execve("/usr/local/bin/drbdsetup-84", ["drbdsetup", "status", "--json"], 0x7ffd6a9cc988 /* 40 vars */) = -1 ENOENT (No such file or directory)
execve("/usr/bin/drbdsetup-84", ["drbdsetup", "status", "--json"], 0x7ffd6a9cc988 /* 40 vars */) = -1 ENOENT (No such file or directory)
execve("/bin/drbdsetup-84", ["drbdsetup", "status", "--json"], 0x7ffd6a9cc988 /* 40 vars */) = -1 ENOENT (No such file or directory)
execve("/usr/local/games/drbdsetup-84", ["drbdsetup", "status", "--json"], 0x7ffd6a9cc988 /* 40 vars */) = -1 ENOENT (No such file or directory)
execve("/usr/games/drbdsetup-84", ["drbdsetup", "status", "--json"], 0x7ffd6a9cc988 /* 40 vars */) = -1 ENOENT (No such file or directory)
execve("/lib/drbd/drbdsetup-84", ["drbdsetup", "status", "--json"], 0x7ffd6a9cc988 /* 40 vars */) = 0
status: unrecognized option '--json'
```

Are there plans to support 8.4 or is only DRBD 9 supported?

**Comment by MalloZup** [comment]
@vvidic thx for issue I will look at

**Comment by MalloZup** [comment]
@vvidic  ok I have checked this. I'm tagging this issue as open question with all maintainer.

@stefanotorresi and others we might need to discuss if we want to support a lower version of DRBD 8.4 or we might need to update documentation to state we need to say only drbd 9 supported

I personally don't know the 8.4 version but we might need to check if the API and data are the same among the 2 version. this could require some efforts.

**Comment by vvidic** [comment]
On Mon, Dec 30, 2019 at 02:04:36AM -0800, Dario Maiocchi wrote:
> @vvidic  ok I have checked this. I'm tagging this issue as open question with all maintainer.
> 
> @stefanotorresi and others we might need to discuss if we want to support a lower version of DRBD 8.4 or we might need to update documentation to state we need to say only drbd 9 supported
> 
> I personally don't know the 8.4 version but we might need to check if the API and data are the same among the 2 version. this could require some efforts.

Right, this is what the drbdsetup command gives for 8.4:

# drbdsetup status --verbose --statistics
drbddemo role:Primary suspended:no
    write-ordering:flush
  volume:0 minor:1 disk:UpToDate
      size:102360 read:0 written:0 al-writes:8 bm-writes:0 upper-pending:0 lower-pending:0 al-suspended:yes
      blocked:no
  peer connection:Connecting role:Unknown congested:no
    volume:0 replication:Off peer-disk:DUnknown resync-suspended:no
        received:0 sent:0 out-of-sync:102360 pending:0 unacked:0

It might be just enough to report the same metrics as DRBD 9.

-- 
Valentin

**Comment by stefanotorresi** [comment]
this might require significant effort, both to implement and to maintain; we need to assess if supporting older versions of DRBD is desirable/required.

**Comment by vvidic** [comment]
Yes, I understand. It's just that DRBD 9 is not in the mainline kernel, so some percent of the people will still run 8.4. If I have time, perhaps I can try to propose a patch for parsing 8.4 output?

**Comment by stefanotorresi** [comment]
thanks @vvidic! We'll discuss this internally and let you know shortly if this is something we want to add; I wouldn't want you to spend time in a PR that I'm not sure would end up being merged at this time.

**Comment by stefanotorresi** [comment]
Hey @vvidic, so we thought about this a little, and I regret to inform you that we eventually decided not to move on with the support for DRBD 8.4.
This project is developed mainly by SUSE employees for usage in SUSE Linux Enterprise Server, which ships DRBD version 9.0: we have no real use case for supporting older versions.
We could have welcomed the addition you suggested, but we found out that DBRD 8.4 is actually already covered by the [Prometheus Node Exporter](https://github.com/prometheus/node_exporter) project, with a [dedicated collector that scans the proc filesystem for stats](https://github.com/prometheus/node_exporter/blob/master/collector/drbd_linux.go).
I hope that satisfies your needs, and I'm sorry we can't do much else to help you at the moment!
